compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -522,18 +525,42 @@ data DependencyInfo = DependencyInfo
   -- flags which define which packages are exposed by default.
   }
 
+showDeps :: [((LF.PackageId, UnitId), LF.Version)] -> String
+showDeps deps =
+    intercalate
+        ", "
+        [ T.unpack (LF.unPackageId pkgId) <> " (" <> unitIdString unitId <> "): " <>
+        DA.Pretty.renderPretty ver
+        | ((pkgId, unitId), ver) <- deps
+        ]
+
+checkForInconsistentLfVersions :: LF.Version -> DependencyInfo -> Either String ()
+checkForInconsistentLfVersions lfTarget DependencyInfo{dalfsFromDependencies, mainUnitIds}
+  | null inconsistentLfDeps = Right ()
+  | otherwise = Left $ concat
+        [ "Targeted LF version "
+        , DA.Pretty.renderPretty lfTarget
+        , " but dependencies have different LF versions: "
+        , showDeps inconsistentLfDeps
+        ]
+  where
+    inconsistentLfDeps =
+        filter (\(_, ver) -> ver /= lfTarget) $
+            [ ( (LF.dalfPackageId decodedDalfPkg, decodedUnitId)
+              , (LF.packageLfVersion . LF.extPackagePkg . LF.dalfPackagePkg) decodedDalfPkg
+              )
+            | DecodedDalf{..} <- dalfsFromDependencies
+            , decodedUnitId `elem` mainUnitIds
```

Comment:
Let’s turn this into a `Set`. The linear lookup seems like we’re just asking to end up on accidentallyquadratic :)

Reply:
You want to turn the `mainUnitIds` field into a `Set` or just here for the lookup?